src/System.Text.Json/tests/Serialization/Object.WriteTests.cs

```diff
@@ -75,5 +75,12 @@ private class Indexer
 
             public string NonIndexerProp { get; set; }
         }
+
+        [Fact]
+        public static void WriteCollectionAsObject()
+        {
+            string json = JsonSerializer.Serialize(new { Prop = (object)new[] { 0 } });
+            Assert.Equal(@"{""Prop"":[0]}", json);
+        }
     }
 }
```

Comment:
Please add tests with different properties with the same `Type` but with different attributes (e.g. `[JsonIgnore]`) to ensure the cache's key works as expected.